content/blog/2024-08-01-gsoc-multiboot2.mdx

```diff
@@ -0,0 +1,98 @@
+Without this tag, the bootloader, GRUB in this case, continues to search for it indefinitely, leading to the infinite loop.
+Now that the issue was resolved, a new contender emerged: the "out of memory" error.
+Let the games begin!
+
+### The "Out of Memory" Error
+
+The "out of memory" error was a puzzling one.
+The first instinct was to increase the memory size, but no matter how much memory was allocated, the error persisted.
+This meant that either something was null or the area where the memory was being allocated was already populated due to possible fragmentation.
+
+Since the error was occuring in the [`grub_relocator_alloc_chunk_align`](https://elixir.bootlin.com/grub/grub-2.06/source/grub-core/lib/relocator.c#L1371) function, I decided to find its caller to reconstitute the flow and logic leading here.
+There were several difficulties with debugging this.
+
+Firstly, I was connected to a 64-bit `gdb` server, but the error was occurring in 32-bit mode so the registers were all messed up.
+To make life a bit easier, I ended up using `qemu-system-i386` instead of `qemu-system-x86_64`.
+Sure, I couldn't boot Unikraft like this, but I could at least debug the issue at hand.
+
+Additionally, I had no symbols.
+Therefore, I had to load them manually using `add-symbol-file` in `gdb` for each module I was interested in.
+
+It is important to mention what I mean by "module".
+GRUB has an interesting way of organizing its modules and object files.
+This might be misleading, since the extensions are `.mod` and `.module`.
+In short, `.mod` files are the final binary that will be loaded at boot time, while `.module`s are intermediate unstripped object files that contains symbol and debug information.
+For the purpose of debugging, I needed the `.module` files.
+
+Obviously, the first module I looked into was `relocator.module`.
+I then needed to determine the base address where I had added my breakpoint.
+The plan was to:
+
+- print the address of the function using [`__builtin_return_address(0)`](https://gcc.gnu.org/onlinedocs/gcc/Return-Address.html#index-_005f_005fbuiltin_005freturn_005faddress)
+- object dump the module to find the offset of the function
+- get the base address by substracting the offset from the address
+
+Pretty straightforward, right?
+Normally, yes.
+But here?
+Well, not quite.
+After performing this process with the printed address, `gdb` was showing code from [`malloc_in_range`](https://elixir.bootlin.com/grub/grub-2.06/source/grub-core/lib/relocator.c#L416) instead of [`grub_relocator_alloc_chunk_align`](https://elixir.bootlin.com/grub/grub-2.06/source/grub-core/lib/relocator.c#L1371).
+Additionally, it did not match the assembly, which corresponded to the manual breakpoint I had added: a dummy volatile integer set to zero and an infinite loop waiting for a change in value (once in `gdb`, I set the register value to 1 to break the loop).
```

Comment:
Uhh, this is fine, but I was more referring to like a small C snippet I guess :sweat_smile:. I would not understand what you said here without context for example :stuck_out_tongue:. Maybe, besides this explanation, you could also exemplify through a snippet
```C
volatile int dummy_var = 1;
while (dummy_var);
```
Or however else you used it :smile: